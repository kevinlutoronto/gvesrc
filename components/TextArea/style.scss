@import '../../styles/base';

.wrapper {
  @extend %input--wrapper;
  height: auto;
  display: flex;
  flex-direction: row;
  &:hover {
    @extend %input--wrapper--hover;
  }
}

.wrapperLarge {
  @extend %input--wrapper--large;
}

.wrapperFocused {
  @extend %input--wrapper--focus;
}

.wrapperDisabled {
  @extend %input--wrapper--disabled;
}

.wrapperError {
  @extend %input--wrapper--error;
  &:hover {
    @extend %input--wrapper--error;
  }
}

.textarea {
  @extend %input;
  // need to override padding here because of autosizing
  padding: 10px;
  @extend %type--body;
  // box-sizing: content-box is required for autosizing
  box-sizing: content-box;
  border: 0;
  outline: none;
  width: 100%;
  order: 1;
  resize: none;
  padding: 10px;
}

.large {
  @extend %input--large;
}

.disabled {
  @extend %input--disabled;
}
