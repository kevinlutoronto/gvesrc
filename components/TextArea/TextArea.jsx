// @flow
import { toUpper } from 'lodash';
import React, { Component } from 'react';
import cx from 'classnames';
import { getDataAttrs } from '../../dataUtils';
import type { Data } from '../../types';
import Autogrow from './autogrow';
import style from './style.scss';

type Props = {|
  autosize: boolean,
  disabled: boolean,
  error: boolean,
  maxLength: string | number,
  onBlur: () => void,
  onChange: () => void,
  onFocus: () => void,
  placeholder: string,
  rows: string,
  size: string,
  value: string,
  data?: Data,
|};

class TextArea extends Component<Props> {
  static defaultProps = {
    placeholder: '',
    rows: '3',
    autosize: true,
  };

  autogrow: Object;

  componentWillUnmount() {
    this.autogrow && this.autogrow.destroy();
  }

  onRef = (el: ?Object) => {
    if (el && this.props.autosize) {
      this.autogrow = new Autogrow(el);
    }
  };

  render() {
    const {
      data,
      disabled,
      error,
      maxLength,
      onBlur,
      onChange,
      onFocus,
      placeholder,
      rows,
      size,
      value,
    } = this.props;

    const wrapperStyle = cx(style.wrapper, {
      [style.wrapperDisabled]: disabled,
      [style.wrapperError]: error,
      [style.wrapperLarge]: toUpper(size) === 'LARGE',
    });

    const textareaStyle = cx(style.textarea, {
      [style.disabled]: disabled,
      [style.large]: toUpper(size) === 'LARGE',
    });

    return (
      <div className={wrapperStyle} {...getDataAttrs(data)}>
        <textarea
          ref={this.onRef}
          className={textareaStyle}
          disabled={disabled}
          onChange={onChange}
          rows={rows}
          onFocus={(e: { target: { parentElement: { className: string } } }) => {
            /* eslint-disable no-param-reassign */
            e.target.parentElement.className += ` ${style.wrapperFocused}`;
            /* eslint-enable no-param-reassign */
            onFocus && onFocus();
          }}
          onBlur={(e: { target: { parentElement: { classList: Object } } }) => {
            e.target.parentElement.classList.remove(style.wrapperFocused);
            onBlur && onBlur();
          }}
          placeholder={placeholder}
          value={value}
          maxLength={maxLength}
        />
      </div>
    );
  }
}

TextArea.displayName = 'Plasma@TextArea';

export default TextArea;
