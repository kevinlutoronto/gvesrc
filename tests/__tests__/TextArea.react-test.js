/* eslint-disable */

import React from 'react';
import TextArea from '../../components/TextArea/TextArea';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';

describe('TextArea Component', () => {
  test('render', () => {
    const component = renderer.create(<TextArea />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('focus', () => {
    const onFocus = jest.fn();
    const component = mount(<TextArea onFocus={onFocus} />);
    component.find('textarea').simulate('focus');
    expect(onFocus).toHaveBeenCalledTimes(1);
  });

  test('blur', () => {
    const onFocus = jest.fn();
    const onBlur = jest.fn();
    const component = mount(<TextArea onFocus={onFocus} onBlur={onBlur} />);
    component.find('textarea').simulate('blur');
    expect(onBlur).toHaveBeenCalledTimes(1);
  });
});
