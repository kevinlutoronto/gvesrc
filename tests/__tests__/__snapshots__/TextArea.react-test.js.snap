// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea Component render 1`] = `
<div
  className="wrapper"
>
  <textarea
    className="textarea"
    disabled={undefined}
    maxLength={undefined}
    onBlur={[Function]}
    onChange={undefined}
    onFocus={[Function]}
    placeholder=""
    rows="3"
    value={undefined}
  />
</div>
`;
